src/basic/fileio.c

```diff
@@ -259,7 +259,14 @@ int read_full_stream(FILE *f, char **contents, size_t *size) {
                 if (ferror(f))
                         return -errno;
 
-                if (feof(f))
+                /* stdio will only set EOF after attempting to read past the
+                 * end of the file.  In the common regular file case where
+                 * know, allocate, and read the file's exact size, EOF isn't
+                 * found yet.  Do our own check against size here to avoid
+                 * doubling the buffer just for fread() to read nothing and
+                 * find eof on the next iteration.
+                 */
+                if (feof(f) || (st_size > 0 && st_size == l))
```

Comment:
well, i'd always focus on minimizing the nr of actual syscalls generated. Hence I'd really read the additional byte right away, instead of doing that in a second read. After all memory is cheap, syscalls are more expensive.